pkg/chains/storage/oci/attestation.go

```diff
@@ -0,0 +1,100 @@
+// limitations under the License.
+
+package oci
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/pkg/errors"
+	"github.com/sigstore/cosign/v2/pkg/oci/mutate"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/cosign/v2/pkg/oci/static"
+	"github.com/sigstore/cosign/v2/pkg/types"
+	"github.com/tektoncd/chains/pkg/chains/storage/api"
+	"github.com/tektoncd/chains/pkg/config"
+	"knative.dev/pkg/logging"
+)
+
+var (
+	_ api.Storer[name.Digest, in_toto.Statement] = &AttestationStorer{}
+)
+
+// AttestationStorer stores in-toto Attestation payloads in OCI registries.
+type AttestationStorer struct {
+	// repo configures the repo where data should be stored.
+	// If empty, the repo is inferred from the Artifact.
+	repo *name.Repository
+	// remoteOpts are additional remote options (i.e. auth) to use for client operations.
+	remoteOpts []remote.Option
+}
+
+func NewAttestationStorer(cfg config.Config) (*AttestationStorer, error) {
+	var repo *name.Repository
+	if r := cfg.Storage.OCI.Repository; r != "" {
+		resolved, err := name.NewRepository(r)
+		if err != nil {
+			return nil, errors.Wrapf(err, "%s is not a valid repository", r)
+		}
+		repo = &resolved
+	}
+
+	return &AttestationStorer{
+		repo: repo,
+	}, nil
+}
+
+func (s *AttestationStorer) Store(ctx context.Context, req *api.StoreRequest[name.Digest, in_toto.Statement]) (*api.StoreResponse, error) {
+	logger := logging.FromContext(ctx)
+	// upload an attestation for each subject
+	logger.Info("Starting to upload attestations to OCI ...")
+	for _, subj := range req.Payload.Subject {
+		imageName := fmt.Sprintf("%s@sha256:%s", subj.Name, subj.Digest["sha256"])
+		logger.Infof("Starting attestation upload to OCI for %s...", imageName)
+
+		repo := req.Artifact.Repository
+		if s.repo != nil {
+			repo = *s.repo
+		}
+		se, err := ociremote.SignedEntity(req.Artifact, ociremote.WithRemoteOptions(s.remoteOpts...))
+		if err != nil {
+			return nil, errors.Wrap(err, "getting signed image")
+		}
+
+		// Create the new attestation for this entity.
+		attOpts := []static.Option{static.WithLayerMediaType(types.DssePayloadType)}
+		if req.Bundle.Cert != nil {
+			attOpts = append(attOpts, static.WithCertChain(req.Bundle.Cert, req.Bundle.Chain))
+		}
+		att, err := static.NewAttestation(req.Bundle.Signature, attOpts...)
+		if err != nil {
+			return nil, err
+		}
+		newImage, err := mutate.AttachAttestationToEntity(se, att)
+		if err != nil {
+			return nil, err
+		}
+
+		// Publish the signatures associated with this entity
+		if err := ociremote.WriteAttestations(repo, newImage, ociremote.WithRemoteOptions(s.remoteOpts...)); err != nil {
+			return nil, err
+		}
+		logger.Infof("Successfully uploaded attestation for %s", imageName)
+	}
+	return &api.StoreResponse{}, nil
```

Comment:
`return nil, nil` ?

Reply:
I thought about this - this is inspired by gRPC which prefers this style of successful responses (it just so happens that there's no content).

The difference here is negligible though so 🤷